src/webgpu/shader/validation/expression/call/builtin/atomics.spec.ts

```diff
@@ -0,0 +1,66 @@
+export const description = `
+Validation tests for atomic builtins.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const atomicOps = [
+  'atomicAdd(&a, 1)',
+  'atomicSub(&a, 1)',
+  'atomicMax(&a, 1)',
+  'atomicMin(&a, 1)',
+  'atomicAnd(&a, 1)',
+  'atomicOr(&a, 1)',
+  'atomicXor(&a, 1)',
+  'atomicLoad(&a)',
+  'atomicStore(&a, 1)',
+];
+
+g.test('stage')
+  .specURL('https://www.w3.org/TR/WGSL/#atomic-rmw')
+  .desc(
+    `
+Atomic built-in functions must not be used in a vertex shader stage.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', ['fragment', 'vertex', 'compute'] as const) //
+      .combine('atomicOp', atomicOps)
```

Comment:
We've been moving away from doing it this was as it will end up with code in the URL for the test. You'll want to do something like what is done in https://github.com/gpuweb/cts/blob/main/src/webgpu/shader/validation/parse/align.spec.ts where the array above becomes a hash:

```
const kAtomicOps = {
  add: {
    src: 'atomicAdd(&a, 1)',
  },
  sub: {
    src: 'atomicSub(&a, 1)',
  },
   ...
};
```

And then with the `combine` you do:

```
.combine('atomicOp', keysOf(kAtomicOps)))
```

You'll then access the `kAtomicOps[t.params.atomicOp].src` to get the source to use in the test.

Reply:
Done. thanks!